src/Development/IDE/Core/FileExists.hs

```diff
@@ -0,0 +1,172 @@
+
+-- | A wrapper around a mutable 'FileExistsMap'
+newtype FileExistsMapVar = FileExistsMapVar (Var FileExistsMap)
+
+instance IsIdeGlobal FileExistsMapVar
+
+-- | Grab the current global value of 'FileExistsMap' without acquiring a dependency
+getFileExistsMapUntracked :: Action FileExistsMap
+getFileExistsMapUntracked = do
+  FileExistsMapVar v <- getIdeGlobalAction
+  liftIO $ readVar v
+
+-- | Modify the global store of file exists
+modifyFileExistsAction :: (FileExistsMap -> IO FileExistsMap) -> Action ()
+modifyFileExistsAction f = do
+  FileExistsMapVar var <- getIdeGlobalAction
+  liftIO $ modifyVar_ var f
+
+-- | Modify the global store of file exists
+modifyFileExists :: IdeState -> [(NormalizedFilePath, Bool)] -> IO ()
+modifyFileExists state changes = do
+  FileExistsMapVar var <- getIdeGlobalState state
+  changesMap           <- evaluate $ Map.fromList changes
+
+  -- Mask to ensure that the previous values are flushed together with the map update
+  mask $ \_ -> do
+    modifyVar_ var $ evaluate . Map.union changesMap
+    let flushPreviousValues = do
+          ShakeExtras { state } <- getShakeExtras
+          liftIO $ mapM_ (deleteValue state GetFileExists . fst) changes
+    void $ shakeRun state [flushPreviousValues]
+
+-------------------------------------------------------------------------------------
+
+type instance RuleResult GetFileExists = Bool
+
+data GetFileExists = GetFileExists
+    deriving (Eq, Show, Typeable, Generic)
+
+instance NFData   GetFileExists
+instance Hashable GetFileExists
+instance Binary   GetFileExists
+
+-- | Returns True if the file exists
+getFileExists :: NormalizedFilePath -> Action Bool
+getFileExists fp = use_ GetFileExists fp
+
+-- | Installs the 'getFileExists' rules.
+--   Provides a fast implementation if client supports dynamic watched files.
+--   Creates a global state as a side effect in that case.
+fileExistsRules :: IO LspId -> ClientCapabilities -> VFSHandle -> Rules ()
+fileExistsRules getLspId ClientCapabilities{_workspace}
+  | Just WorkspaceClientCapabilities{_didChangeWatchedFiles} <- _workspace
+  , Just DidChangeWatchedFilesClientCapabilities{_dynamicRegistration} <- _didChangeWatchedFiles
+  , Just True <- _dynamicRegistration
+  = fileExistsRulesFast getLspId
+  | otherwise = fileExistsRulesSlow
+
+--   Requires an lsp client that provides WatchedFiles notifications.
+fileExistsRulesFast :: IO LspId -> VFSHandle -> Rules ()
+fileExistsRulesFast getLspId vfs = do
+  addIdeGlobal . FileExistsMapVar =<< liftIO (newVar [])
+  defineEarlyCutoff $ \GetFileExists file -> do
+    fileExistsMap <- getFileExistsMapUntracked
+    let mbFilesWatched = Map.lookup file fileExistsMap
+    case mbFilesWatched of
+      Just fv -> pure (createKey fv, ([], Just fv))
+      Nothing -> do
+        exist                   <- liftIO $ getFileExistsVFS vfs file
+        ShakeExtras { eventer } <- getShakeExtras
+
+        -- add a listener for VFS Create/Delete file events,
```

Comment:
I don’t think you are preventing any race conditions here. `addListener` adds the listener asynchronously by going through the `eventer` which will basically just do a `writeChan`.

Also if the request to adds the listener fails for whatever reason, you will never notice that. The proper fix for that would probably be to track the request and wait to see if you get a successful response. If it fails, you have to rerun and at least in theory you also have to rerun once after the listener is registered to prevent events between the initial query and the listener being registered from getting lost.

I would suggest to not try to attack this in this PR given that I don’t see why this request should ever fail (ignoring bugs like invalid messages) and the race seems fairly harmless but we should update the comment.

Reply:
Even if `eventer` is asyncrhonous, the race is prevented by the key inserted in the `fileExistsMap`. But if the request fails we'll never notice, agreed.

Reply:
the key is inserted once you receive an event, not once the listener is registered. So I don’t think it prevents you from registering the same listener multiple times.

Reply:
The code was wrong from the very first version. It was not inserting anything at all in the map. This would result in unnecessary calls to `getFileExistsVFS`. I have fixed it to insert in the map and now it will only ever set up the listener once, iiuc. Recovering from a failed listener install seems more tricky:
- if it failed with an exception, then no insertion would ever happen and we'd be good. But this is not the case
- as it fails asynchronously, we would need a continuation to remove the entry from `getFileExistsMap`. But as far as I can see there's no way to install such continuation.